datetime tests out of range macOS & python 3.12

https://github.com/PyO3/pyo3/actions/runs/8118596439/job/22193138483?pr=3917#step:22:268a

I haven't tried to reproduce yet, if it does reproduce locally then it should be reasonably straightforward to work out how to adjust the test to stay in bounds. 

Fixing this would help reduce CI flakiness!

I've marked this good first issue as it might be a nice way for someone to familiarise with the codebase. (In particular our python based part of the test suite.)

I don't think CI jobs stay around forever, so here's the error:

```python
  _________________________ test_datetime_from_timestamp _________________________
  
      @given(dt=st.datetimes(MIN_DATETIME, MAX_DATETIME))
  >   @example(dt=pdt.datetime(1971, 1, 2, 0, 0))
  
  tests/test_datetime.py:230: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  dt = datetime.datetime(9999, 12, 31, 0, 0, fold=1)
  
      @given(dt=st.datetimes(MIN_DATETIME, MAX_DATETIME))
      @example(dt=pdt.datetime(1971, 1, 2, 0, 0))
      def test_datetime_from_timestamp(dt):
          if PYPY and dt < pdt.datetime(1900, 1, 1):
              pytest.xfail("pdt.datetime.timestamp will raise on PyPy with dates before 1900")
      
  >       ts = pdt.datetime.timestamp(dt)
  E       ValueError: year 10000 is out of range
  E       Falsifying example: test_datetime_from_timestamp(
  E           dt=datetime.datetime(9999, 12, 31, 0, 0, fold=1),
  E       )
  E       
  E       You can reproduce this example by temporarily adding @reproduce_failure('6.98.15', b'AAAfPwseAAAAAAAAAQ==') as a decorator on your test case
  
  tests/test_datetime.py:235: ValueError
  =========================== short test summary info ============================
  FAILED tests/test_datetime.py::test_datetime_from_timestamp - ValueError: year 10000 is out of range
  Falsifying example: test_datetime_from_timestamp(
      dt=datetime.datetime(9999, 12, 31, 0, 0, fold=1),
  )
  
  You can reproduce this example by temporarily adding @reproduce_failure('6.98.15', b'AAAfPwseAAAAAAAAAQ==') as a decorator on your test case
```

Looks like windows 3.12 can fail too:

```
 ================================== FAILURES ===================================
  ________________________ test_datetime_from_timestamp _________________________
  
      @given(dt=st.datetimes(MIN_DATETIME, MAX_DATETIME))
  >   @example(dt=pdt.datetime(1971, 1, 2, 0, 0))
  
  tests\test_datetime.py:230: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  
  dt = datetime.datetime(3001, 1, 18, 8, 0, fold=1)
  
      @given(dt=st.datetimes(MIN_DATETIME, MAX_DATETIME))
      @example(dt=pdt.datetime(1971, 1, 2, 0, 0))
      def test_datetime_from_timestamp(dt):
          if PYPY and dt < pdt.datetime(1900, 1, 1):
              pytest.xfail("pdt.datetime.timestamp will raise on PyPy with dates before 1900")
      
  >       ts = pdt.datetime.timestamp(dt)
  E       OSError: [Errno 22] Invalid argument
  E       Falsifying example: test_datetime_from_timestamp(
  E           dt=datetime.datetime(3001, 1, 18, 8, 0, fold=1),
  E       )
  E       
  E       You can reproduce this example by temporarily adding @reproduce_failure('6.100.1', b'AAAD6QARCAAAAAAAAQ==') as a decorator on your test case
  
  tests\test_datetime.py:235: OSError
  =========================== short test summary info ===========================
  FAILED tests/test_datetime.py::test_datetime_from_timestamp - OSError: [Errno 22] Invalid argument
  Falsifying example: test_datetime_from_timestamp(
      dt=datetime.datetime(3001, 1, 18, 8, 0, fold=1),
  )
  
  You can reproduce this example by temporarily adding @reproduce_failure('6.100.1', b'AAAD6QARCAAAAAAAAQ==') as a decorator on your test case
  ======================== 1 failed, 147 passed in 3.87s ========================
```

--- a/pytests/tests/test_datetime.py
+++ b/pytests/tests/test_datetime.py
@@ -6,7 +6,7 @@ import sys
 
 import pyo3_pytests.datetime as rdt
 import pytest
-from hypothesis import example, given
+from hypothesis import given
 from hypothesis import strategies as st
 
 
@@ -227,7 +227,6 @@ def test_datetime_typeerror():
 
 
 @given(dt=st.datetimes(MIN_DATETIME, MAX_DATETIME))
-@example(dt=pdt.datetime(1971, 1, 2, 0, 0))
 def test_datetime_from_timestamp(dt):
     if PYPY and dt < pdt.datetime(1900, 1, 1):
         pytest.xfail("pdt.datetime.timestamp will raise on PyPy with dates before 1900")

--- a/pytests/tests/test_datetime.py
+++ b/pytests/tests/test_datetime.py
@@ -6,7 +6,7 @@ import sys
 
 import pyo3_pytests.datetime as rdt
 import pytest
-from hypothesis import given
+from hypothesis import example, given
 from hypothesis import strategies as st
 
 
@@ -93,15 +93,21 @@ def test_invalid_date_fails():
 
 @given(d=st.dates(MIN_DATETIME.date(), MAX_DATETIME.date()))
 def test_date_from_timestamp(d):
-    if PYPY and d < pdt.date(1900, 1, 1):
-        pytest.xfail("pdt.datetime.timestamp will raise on PyPy with dates before 1900")
+    try:
+        ts = pdt.datetime.timestamp(d)
+    except Exception:
+        # out of range for timestamp
+        return
 
-    ts = pdt.datetime.timestamp(
-        pdt.datetime.combine(d, pdt.time(0)).replace(
-            tzinfo=pdt.timezone(pdt.timedelta(0))
-        )
-    )
-    assert rdt.date_from_timestamp(int(ts)) == pdt.date.fromtimestamp(ts)
+    try:
+        expected = pdt.date.fromtimestamp(ts)
+    except Exception as pdt_fail:
+        # date from timestamp failed; expect the same from Rust binding
+        with pytest.raises(type(pdt_fail)) as exc_info:
+            rdt.date_from_timestamp(ts)
+        assert str(exc_info.value) == str(pdt_fail)
+    else:
+        assert rdt.date_from_timestamp(int(ts)) == expected
 
 
 @pytest.mark.parametrize(
@@ -231,12 +237,23 @@ def test_datetime_typeerror():
 
 
 @given(dt=st.datetimes(MIN_DATETIME, MAX_DATETIME))
+@example(dt=pdt.datetime(1971, 1, 2, 0, 0))
 def test_datetime_from_timestamp(dt):
-    if PYPY and dt < pdt.datetime(1900, 1, 1):
-        pytest.xfail("pdt.datetime.timestamp will raise on PyPy with dates before 1900")
+    try:
+        ts = pdt.datetime.timestamp(dt)
+    except Exception:
+        # out of range for timestamp
+        return
 
-    ts = pdt.datetime.timestamp(dt.replace(tzinfo=pdt.timezone(pdt.timedelta(0))))
-    assert rdt.datetime_from_timestamp(ts) == pdt.datetime.fromtimestamp(ts)
+    try:
+        expected = pdt.datetime.fromtimestamp(ts)
+    except Exception as pdt_fail:
+        # datetime from timestamp failed; expect the same from Rust binding
+        with pytest.raises(type(pdt_fail)) as exc_info:
+            rdt.datetime_from_timestamp(ts)
+        assert str(exc_info.value) == str(pdt_fail)
+    else:
+        assert rdt.datetime_from_timestamp(ts) == expected
 
 
 def test_datetime_from_timestamp_tzinfo():

--- a/pytests/tests/test_datetime.py
+++ b/pytests/tests/test_datetime.py
@@ -56,11 +56,11 @@ else:
 IS_WINDOWS = sys.platform == "win32"
 
 if IS_WINDOWS:
-    MIN_DATETIME = pdt.datetime(1971, 1, 2, 0, 0)
+    MIN_DATETIME = pdt.datetime(1970, 1, 1, 0, 0, 0)
     if IS_32_BIT:
-        MAX_DATETIME = pdt.datetime(3001, 1, 19, 4, 59, 59)
+        MAX_DATETIME = pdt.datetime(2038, 1, 18, 23, 59, 59)
     else:
-        MAX_DATETIME = pdt.datetime(3001, 1, 19, 7, 59, 59)
+        MAX_DATETIME = pdt.datetime(3000, 12, 31, 23, 59, 59)
 else:
     if IS_32_BIT:
         # TS ±2147483648 (2**31)

--- a/pytests/tests/test_datetime.py
+++ b/pytests/tests/test_datetime.py
@@ -96,7 +96,11 @@ def test_date_from_timestamp(d):
     if PYPY and d < pdt.date(1900, 1, 1):
         pytest.xfail("pdt.datetime.timestamp will raise on PyPy with dates before 1900")
 
-    ts = pdt.datetime.timestamp(pdt.datetime.combine(d, pdt.time(0)))
+    ts = pdt.datetime.timestamp(
+        pdt.datetime.combine(d, pdt.time(0)).replace(
+            tzinfo=pdt.timezone(pdt.timedelta(0))
+        )
+    )
     assert rdt.date_from_timestamp(int(ts)) == pdt.date.fromtimestamp(ts)
 
 
@@ -231,7 +235,7 @@ def test_datetime_from_timestamp(dt):
     if PYPY and dt < pdt.datetime(1900, 1, 1):
         pytest.xfail("pdt.datetime.timestamp will raise on PyPy with dates before 1900")
 
-    ts = pdt.datetime.timestamp(dt)
+    ts = pdt.datetime.timestamp(dt.replace(tzinfo=pdt.timezone(pdt.timedelta(0))))
     assert rdt.datetime_from_timestamp(ts) == pdt.datetime.fromtimestamp(ts)
 
 

